include/bk/lexer.h

```diff
@@ -0,0 +1,54 @@
+/*
+   Copyright 2019 Simon (psyomn) Symeonidis
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+#pragma once
+
+#include <bk/token.h>
+#include <bk/module.h>
+
+#include <iostream>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace bk {
+struct Lexer {
+public:
+
+  Lexer(const std::string&);
+  ~Lexer();
+
+  void process(); // TODO maybe this should be moved to a util class
+  void tokenizeFile(const std::string&);
+  void clearModules() {
+    mModules.clear();
+  }
+
+  inline std::size_t numModules() {
+    return mModules.size();
+  }
+
+private:
+
+  std::size_t             mLineNum;
+  std::string             mRoot;
+  std::vector<std::string>mFiles;
+  std::string             mAcceptExtension;
+  std::vector<bk::Module> mModules;
+
+  friend std ::ostream& operator<<(std::ostream&,
```

Comment:
I would avoid using friend function here. In this case, you can just add the operator directly in your class. Using `friend` for an external operator is only useful if you want your class to be the right hand of the operator and still need access to implementation details of the class.

Reply:
So I'm not sure if I can actually add these within the class without friend. I've incorporated your feedback. Thanks!